### PR TITLE
Include LinalgStructuredOps headers from LLVM

### PR DESCRIPTION
**Pull request by pzread**
This change tries to import the `LinalgStructuredOps.cpp.inc` from LLVM instead of keeping a copy in IREE. It will help us avoid #9036 in the future.

There is a compatibility issue between IREE and LLVM cmake system. The tablegen libraries in LLVM are only defined as custom targets and can not be used in `target_link_libraries`, so adding those targets to the `DEPS` of `iree_cc_library` breaks the build as the rule tries to link them.

I'm not sure about the best solution for this. In IREE, we can use `add_dependences` to add LLVM tablegen libraries but it will break the mapping between bazel and cmake, makes it harder to maintain. Another option is defining those targets as [Interface Libraries](https://cmake.org/cmake/help/latest/command/add_library.html#interface-libraries) in the LLVM upstream, but that might need to change LLVM a lot.

My current solution is to re-export those tablegen targets as interface libraries in IREE, so we can still map from bazel to cmake and link those targets as normal libraries in IREE.

### CODE REVIEW

**Review by MaheshRavishankar — APPROVED** [review]
I am happy with the code changes. (Thanks a lot for this, cause last time I tried this I had bazel build issues that seems to have been addressed here). LGTM but would be good to wait for someone more familiar with the build dependencies to also stamp.

File: build_tools/third_party/llvm-project/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+# SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+###############################################################################
+# Re-export LLVM header-only targets as header-only libraries.                #
+###############################################################################
+
+add_library(MLIRLinalgStructuredOpsIncGenLib INTERFACE)
+add_dependencies(MLIRLinalgStructuredOpsIncGenLib
+  MLIRLinalgStructuredOpsIncGen
+)
```
Comment (ScottTodd) [inline build_tools/third_party/llvm-project/CMakeLists.txt:14]: This approach seems reasonable enough to me at this scale. If we end up with 10s of these across the project then we'd want to be more careful.

I'm a bit foggy on the specifics right now, but here's another general idea to consider: we already have [`iree_compiler_cc_library`](https://github.com/google/iree/blob/18b9df4ac41f7cbf86ec0e036771b39f9f3420f4/iree/build_defs.oss.bzl#L49-L60) in Bazel [getting converted](https://github.com/google/iree/blob/18b9df4ac41f7cbf86ec0e036771b39f9f3420f4/build_tools/bazel_to_cmake/bazel_to_cmake_converter.py#L335-L336) to [`iree_cc_library`](https://github.com/google/iree/blob/main/build_tools/cmake/iree_cc_library.cmake) in CMake. We could add another argument to `iree_compiler_cc_library` (Bazel) that would get converted to a new argument on `iree_cc_library` (CMake) that would use `add_dependences`.

Might want Geoffrey to weigh in when he gets back :)

**Comment by pzread** [comment]
@GMNGeoffrey Ping : )

File: build_tools/third_party/llvm-project/CMakeLists.txt (hunk 1 above)
Comment (GMNGeoffrey) [reply build_tools/third_party/llvm-project/CMakeLists.txt:14]: I didn't know about `iree_compiler_cc_library`. Looks like it's new. Your idea seems good Scott, but the approach in this PR also seems like an improvement. How about we merge this and then try out the `iree_compiler_cc_library` idea?

**Review by GMNGeoffrey — APPROVED** [review]

**Comment by GMNGeoffrey** [comment]
> @GMNGeoffrey Ping : )

Thanks for the ping. I missed this in the sea of post-vacation email

File: build_tools/third_party/llvm-project/CMakeLists.txt (hunk 1 above)
Comment (pzread) [reply build_tools/third_party/llvm-project/CMakeLists.txt:14]: Filed #9146 to track this.

**Pull request merged**